### DISCUSSION TITLE
Replace Dynamo with @aws-sdk/lib-dynamodb/DynamoDBDocumentClient for v3?

### DISCUSSION THREAD

**Discussion by tamakisquare** [sensedeep/dynamodb-onetable#68] [Q&A] [answered]
A quick skim through both pieces of code tells me that they are essentially the same. 

Is there a plan to replace Dynamo with @aws-sdk/lib-dynamodb/DynamoDBDocumentClient (for v3)?

**Comment by mobsense** [comment, marked as answer]
We created the AWS V3 support before V3 added (beta / prototype) document client support.

The Dynamo code is probably faster as it does not try to do the whole DocumentClient API.

Is there a benefit in replacing it when the AWS SDK V3 DocumentClient is production ready?

**Reply by mobsense** [reply]
A bit more data. DocumentClient doesn't have complete support for mapping Sets into and out of Javascript in V2. 

Unclear if the V3 DocumentClient suffers similarly.

**Reply by tamakisquare** [reply]
V3 document client (@aws-sdk/lib-dynamodb/DynamoDBDocumentClient) provides two-way bridge between JavaScript types and DynamoDB AttributeValues as explained in the following excerpt taken from [their in-code documentation](https://github.com/aws/aws-sdk-js-v3/blob/a51c2f4239bb0722c55503ac8004e150cfdb7043/lib/lib-dynamodb/src/DynamoDBDocumentClient.ts#L74). This is the obvious benefit for OneTable library to adopt DynamoDBDocumentClient, as the work of maintaining data marshalling and unmarshalling can be offloaded to the official library.

>  ## Marshalling Input and Unmarshalling Response Data
>  The document client affords developers the use of native JavaScript types
>  instead of `AttributeValue`s to simplify the JavaScript development
>  experience with Amazon DynamoDB. JavaScript objects passed in as parameters
>  are marshalled into `AttributeValue` shapes required by Amazon DynamoDB.
>  Responses from DynamoDB are unmarshalled into plain JavaScript objects
>  by the `DocumentClient`. The `DocumentClient` does not accept
>  `AttributeValue`s in favor of native JavaScript types.
> 
>  |          JavaScript Type          | DynamoDB AttributeValue |
>  | :-------------------------------: | ----------------------- |
>  |              String               | S                       |
>  |          Number / BigInt          | N                       |
>  |              Boolean              | BOOL                    |
>  |               null                | NULL                    |
>  |               Array               | L                       |
>  |              Object               | M                       |
>  |   Set\<Uint8Array, Blob, ...\>    | BS                      |
>  |       Set\<Number, BigInt\>       | NS                      |
>  |           Set\<String\>           | SS                      |
>  | Uint8Array, Buffer, File, Blob... | B                       |
